src/backend/session.rs

```diff
@@ -0,0 +1,193 @@
+        }
+        if let Some(monitor) = &self.monitor {
+            let _ = monitor.session_closed(handle_token).await;
+        }
+        server.remove::<Self, _>(&self.path).await?;
+        Ok(())
+    }
+
+    #[zbus(signal)]
+    async fn closed(signal_emitter: &SignalEmitter<'_>) -> zbus::Result<()>;
+}
+
+impl Debug for Session {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Session").field("path", &self.path).finish()
+    }
+}
+
+#[async_trait]
+/// A trait that backends that create long-lived sessions should implement
+/// to be notified when a session has been closed.
+pub trait SessionImpl: Send + Sync {
+    async fn session_closed(&self, session_handle_token: HandleToken)
+        -> crate::backend::Result<()>;
+}
+
+#[derive(SerializeDict, Type, Debug)]
+#[zvariant(signature = "dict")]
+pub struct CreateSessionResponse {
+    session_id: HandleToken,
+}
+
+impl CreateSessionResponse {
+    pub fn new(handle_token: HandleToken) -> Self {
+        Self {
+            session_id: handle_token,
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct SessionManager {
+    sessions: Arc<Mutex<HashMap<HandleToken, Session>>>,
+}
+
+impl SessionManager {
+    /// Tests whether a session with the given handle token already exists.
+    pub fn contains(&self, handle_token: &HandleToken) -> bool {
+        #[cfg(feature = "tracing")]
+        tracing::debug!("SessionManager::contains: tracked sessions: {}", &self);
+        let sessions = self.sessions.lock().unwrap();
+        sessions.contains_key(handle_token)
+    }
+
+    /// Add a session to the list of tracked sessions.
+    pub fn add(&mut self, session: Session) -> crate::backend::Result<()> {
+        let mut sessions = self.sessions.lock().unwrap();
+        let handle_token = session.handle_token();
+        if sessions.insert(handle_token.clone(), session).is_some() {
+            Err(self.error_exist(&handle_token))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Remove a session from the list of tracked sessions.
+    pub fn remove(&mut self, handle_token: &HandleToken) -> crate::backend::Result<()> {
+        let mut sessions = self.sessions.lock().unwrap();
+        if sessions.remove(handle_token).is_some() {
+            Ok(())
+        } else {
+            Err(self.error_notfound(handle_token))
+        }
+    }
+
+    pub fn error_exist(&self, handle_token: &HandleToken) -> PortalError {
+        let errormsg = format!("A session with handle `{handle_token}` already exists");
+        #[cfg(feature = "tracing")]
+        tracing::error!("{}", errormsg.as_str());
+        PortalError::Exist(errormsg)
+    }
+
+    pub fn error_notfound(&self, handle_token: &HandleToken) -> PortalError {
+        let errormsg = format!("Unknown session: `{handle_token}`");
+        #[cfg(feature = "tracing")]
+        tracing::error!("{}", errormsg.as_str());
+        PortalError::NotFound(errormsg)
+    }
+}
```

Comment:
I would make those private and add something like try_contains that returns a Err if it doesn't exist.

Reply:
That was my initial preference, but there are cases where we want to error out when a session already exists (in `CreateSession`), and cases where we want to error out when a session *doesn't* exist (basically all other methods that act on a previously created session). So we would need two variants of `try_contains` with an opposite logic, and I don't know how to write this in a way that's not going to be confusing.

Reply:
You can have a try_contains and contains, contains would call try_contains().is_ok() and that is pretty much it, no?

Reply:
I went for your suggestion, removed the extra check for existence in the `add()` method, and expanded a bit the documentation to make things more explicit.